aesara/tensor/subtensor_opt.py

```diff
@@ -1661,3 +1663,97 @@ def local_subtensor_SpecifyShape_lift(fgraph, node):
     if new_obj_arg.ndim == 0:
         return [new_obj_arg]
     return [specify_shape(new_obj_arg, shape_arg[len(indices) :])]
+
+
+@register_specialize
+@local_optimizer([Join])
+def local_join_subtensors(fgraph, node):
+    r"""Simplify contiguous :class:`Subtensor`\s inside a :class:`Join`.
+
+    `join((x[:3], x[3:5]), axis=0) -> x[:5]`
+    """
+    # TODO: Generalize to AdvancedSubtensors
+
+    axis, tensors = node.inputs[0], node.inputs[1:]
+
+    try:
+        axis = get_scalar_constant_value(axis)
+    except NotScalarConstantError:
+        return
+
+    for tensor1_idx, (tensor1, tensor2) in enumerate(zip(tensors[:-1], tensors[1:])):
+        # Check that two consecutive Subtensors are operating on the same base tensor
+        if not (
+            (tensor1.owner is not None and isinstance(tensor1.owner.op, Subtensor))
+            and (tensor2.owner is not None and isinstance(tensor2.owner.op, Subtensor))
+            and (tensor1.owner.inputs[0] is tensor2.owner.inputs[0])
+        ):
+            continue
+
+        # Check that subtensors have consecutive indexes across the join axis
+        idxs_tensor1 = indices_from_subtensor(
+            tensor1.owner.inputs[1:], tensor1.owner.op.idx_list
+        )
+        idxs_tensor2 = indices_from_subtensor(
+            tensor2.owner.inputs[1:], tensor2.owner.op.idx_list
+        )
+        try:
+            idxs_axis_tensor1 = idxs_tensor1[axis]
+            idxs_axis_tensor2 = idxs_tensor2[axis]
+        except IndexError:
+            continue
+        if not (
+            isinstance(idxs_axis_tensor1, slice)
+            and isinstance(idxs_axis_tensor2, slice)
+        ):
+            continue
+        start_tensor1, stop_tensor1, step_tensor1 = (
+            idxs_axis_tensor1.start,
+            idxs_axis_tensor1.stop,
+            idxs_axis_tensor1.step,
+        )
+        start_tensor2, stop_tensor2, step_tensor2 = (
+            idxs_axis_tensor2.start,
+            idxs_axis_tensor2.stop,
+            idxs_axis_tensor2.step,
+        )
+        if not (
+            (stop_tensor1 is not None and start_tensor2 is not None)
+            and (stop_tensor1 is start_tensor2)
+            and (step_tensor1 is step_tensor2)
+        ):
+            continue
+
+        # Check that all other idxs of subtensor are the same
+        def equal_idxs(idx1, idx2):
+            if isinstance(idx1, slice) and isinstance(idx2, slice):
+                return (
+                    (idx1.start is idx2.start)
+                    and (idx1.stop is idx2.stop)
+                    and (idx1.step is idx2.step)
+                )
+            else:
+                return idx1 is idx2
```

Comment:
Why `is`?

Reply:
IIRC, because we wanted to avoid use of `equal_computations` and rely on variable identity / constant merging

Reply:
`==` is the best approach, because it requires fewer redundant comparisons across types that implement `__eq__` properly.

Reply:
Oh, I was under the impression that equality of TensorVariables was always `True` (non-empty)!